libdrgn/program.c

```diff
@@ -433,14 +433,20 @@ drgn_program_set_core_dump(struct drgn_program *prog, const char *path)
 		prog->file_segments[j].file_size = phdr->p_filesz;
 		prog->file_segments[j].fd = prog->core_fd;
 		prog->file_segments[j].eio_is_fault = false;
+		/*
+		 * vmcoreinfo_note is a good proxy for the current core being a
+		 * kernel core and we use it to decide whether to zerofill pages
+		 * between p_filesz and p_memsz. We do this because at this
+		 * point only makecoredump excludes zero pages; userland is
+		 * not affected.
+		 */
+		prog->file_segments[j].zerofill = vmcoreinfo_note != NULL;
```

Comment:
It looks like `/proc/kcore` always sets `filesz` to the same value as `memsz`: https://github.com/torvalds/linux/blob/072e51356cd5a4a1c12c1020bc054c99b98333df/fs/proc/kcore.c#L396

I don't see a reason why `filesz` would ever be less than `memsz` for `/proc/kcore`. So, I'd rather be conservative and exclude it from the zero filling behavior.

```suggestion
		prog->file_segments[j].zerofill = vmcoreinfo_note && !is_proc_kcore;
```

Reply:
I think I addressed everything; let me know how you feel about the new PR.